src/Script/Script.ts

```diff
@@ -75,44 +52,40 @@ export class Script {
     }, 0);
   }
 
-  /**
-   * Marks this script as having been updated. It will be recompiled next time something tries
-   * to exec it.
-   */
-  markUpdated(): void {
-    this.module = null;
-    this.moduleSequenceNumber = ++globalModuleSequenceNumber;
+  /** Invalidates the current script module and related data, e.g. when modifying the file. */
+  invalidateModule(): void {
+    delete this.module;
+    if (this.url) URL.revokeObjectURL(this.url);
+    delete this.url;
+    delete this.ramUsage;
+    delete this.ramUsageEntries;
+    for (const [url, dependency] of this.dependencies) {
+      dependency.dependents.delete(this);
+      this.dependencies.delete(url);
+    }
+    this.dependents.forEach((dependent) => {
+      this.dependents.delete(dependent);
```

Comment:
Same; you can just clear the set after.

Edit: Maybe not, there could be subtle effects due to the recursion. The recursion here makes everything very hard to analyze for correctness.

Reply:
Actually it looks like clearing the set isn't even needed. This will be done automatically when the dependents invalidate their own modules (since the invalidation includes going through all dependencies and clearing self from the dependency's dependents set)